File: ubel/policy.py
"""
policy.py — UBEL threshold-based policy evaluator.

Python port of policy.js.  Zero external dependencies.

Policy file schema (JSON):
{
    "severity_threshold":            "high",   // block this level and above
    "block_unknown_vulnerabilities": true       // whether to block unknowns
}

Severity order (ascending): low → medium → high → critical
"unknown" is governed solely by block_unknown_vulnerabilities.

Infections (MAL-*) are always blocked regardless of any policy setting.
"""

from __future__ import annotations

from typing import Tuple

SEVERITY_ORDER = ["low", "medium", "high", "critical"]


def _meets_threshold(candidate: str, threshold: str) -> bool:
    """Return True if *candidate* severity is >= *threshold* severity."""
    try:
        return SEVERITY_ORDER.index(candidate) >= SEVERITY_ORDER.index(threshold)
    except ValueError:
        return False


def evaluate_policy(report: dict) -> Tuple[bool, str]:
    """
    Evaluate policy against a scan report.
    Returns (allowed: bool, reason: str).

    Mirrors evaluatePolicy() in policy.js exactly.
    """
    stats  = report.get("stats", {})
    policy = report.get("policy")

    if not policy:
        raise RuntimeError("No policy returned by API (fail-closed)")

    # ── 1. Infections: always blocked, no policy toggle ──────────────────────
    if stats.get("total_infections", 0) > 0:
        return False, "Blocked: infections detected (always enforced)"

    # ── 2. Severity threshold ─────────────────────────────────────────────────
    raw_threshold = (policy.get("severity_threshold") or "").lower()
    if raw_threshold and raw_threshold in SEVERITY_ORDER:
        severity_stats = (
            stats.get("vulnerabilities_stats", {})
                 .get("severity", {})
        )
        for level in SEVERITY_ORDER:
            if _meets_threshold(level, raw_threshold) and severity_stats.get(level, 0) > 0:
                return (
                    False,
                    f"Blocked by policy: {level} severity vulnerabilities detected "
                    f"(threshold: {raw_threshold})",
                )

    # ── 3. Unknown vulnerabilities ────────────────────────────────────────────
    if policy.get("block_unknown_vulnerabilities") is True:
        unknown_count = (
            stats.get("vulnerabilities_stats", {})
                 .get("severity", {})
                 .get("unknown", 0)
        )
        if unknown_count > 0:
            return (
                False,
                f"Blocked by policy: {unknown_count} unknown-severity vulnerabilities detected",
            )

    return True, "Policy passed"

File: node/sca/policy.js
/**
 * policy.js — UBEL threshold-based policy evaluator.
 *
 * Policy file schema (JSON):
 * {
 *   "severity_threshold": "high",          // block this level and above
 *   "block_unknown_vulnerabilities": true, // whether to block unknowns
 *   "license_risk_threshold": "none"       // block this license risk level
 *                                          // and above; "none" disables
 *                                          // license-risk blocking (default)
 * }
 *
 * Severity order (ascending): low → medium → high → critical
 * "unknown" is governed solely by block_unknown_vulnerabilities.
 *
 * License risk order (ascending): low → medium → high
 * "none" is not a risk level — it means the gate is off. There is
 * deliberately no "unknown" handling here: license detection can't always
 * resolve a package's terms (missing metadata, unparsable free text), and
 * folding "unknown" into an enforced gate would risk blocking installs
 * over a detection gap rather than an actual legal/compliance finding. If
 * that's ever needed, it should be its own explicit opt-in flag, mirroring
 * block_unknown_vulnerabilities above, not bundled into the threshold.
 *
 * Only populated for `health`-mode scans — see engine.js — so this check
 * is a no-op for `check`/`install` scans regardless of the configured
 * threshold.
 *
 * Infections are always blocked regardless of policy.
 */

const SEVERITY_ORDER = ["low", "medium", "high", "critical"];
const LICENSE_RISK_ORDER = ["low", "medium", "high"];

/**
 * Returns true if `candidate` severity is >= `threshold` severity.
 * Both values must be members of SEVERITY_ORDER.
 */
function meetsThreshold(candidate, threshold) {
  return SEVERITY_ORDER.indexOf(candidate) >= SEVERITY_ORDER.indexOf(threshold);
}

/**
 * Returns true if `candidate` license risk is >= `threshold` license risk.
 * Both values must be members of LICENSE_RISK_ORDER.
 */
function meetsLicenseRiskThreshold(candidate, threshold) {
  return LICENSE_RISK_ORDER.indexOf(candidate) >= LICENSE_RISK_ORDER.indexOf(threshold);
}

/**
 * Evaluate policy against a scan report.
 * Returns [allowed: boolean, reason: string]
 *
 * @param {object} report  - Scan report with .stats and .policy fields.
 */
export function evaluatePolicy(report) {
  const stats  = report.stats  || {};
  const policy = report.policy;

  if (!policy) {
    throw new Error("No policy returned by API (fail-closed)");
  }

  // ── 1. Infections: always blocked, no policy toggle ──────────────────────
  if ((stats.total_infections || 0) > 0) {
    return [false, "Blocked: infections detected (always enforced)"];
  }
  if ((report.secrets?.count || 0) > 0) {
    return [false, "Blocked: secrets detected (always enforced)"];
  }

  // ── 2. Severity threshold ─────────────────────────────────────────────────
  const rawThreshold = (policy.severity_threshold || "").toLowerCase();
  if (rawThreshold && SEVERITY_ORDER.includes(rawThreshold)) {
    const severityStats = stats?.vulnerabilities_stats?.severity || {};

    for (const level of SEVERITY_ORDER) {
      if (meetsThreshold(level, rawThreshold) && (severityStats[level] || 0) > 0) {
        return [
          false,
          `Blocked by policy: ${level} severity vulnerabilities detected ` +
          `(threshold: ${rawThreshold})`,
        ];
      }
    }
  }

  // ── 3. Unknown vulnerabilities ────────────────────────────────────────────
  if (policy.block_unknown_vulnerabilities === true) {
    const unknownCount = stats?.vulnerabilities_stats?.severity?.unknown || 0;
    if (unknownCount > 0) {
      return [false, `Blocked by policy: ${unknownCount} unknown-severity vulnerabilities detected`];
    }
  }

  // ── 4. License risk (health-mode scans only; see engine.js) ──────────────
  const rawLicenseThreshold = (policy.license_risk_threshold || "none").toLowerCase();
  if (rawLicenseThreshold !== "none" && LICENSE_RISK_ORDER.includes(rawLicenseThreshold)) {
    const byRisk = stats?.license_stats?.by_risk || {};

    for (const level of LICENSE_RISK_ORDER) {
      if (meetsLicenseRiskThreshold(level, rawLicenseThreshold) && (byRisk[level] || 0) > 0) {
        return [
          false,
          `Blocked by policy: ${byRisk[level]} package(s) with ${level} license risk detected ` +
          `(threshold: ${rawLicenseThreshold})`,
        ];
      }
    }
  }

  return [true, "Policy passed"];
}